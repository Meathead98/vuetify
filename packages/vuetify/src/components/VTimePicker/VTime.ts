// Utils
import pad from '../VDatePicker/util/pad'

// Types
import Vue, { VNode } from 'vue'
import { PropValidator } from 'vue/types/options'

export const enum SelectMode {
  Hour = 'hour',
  Minute = 'minute',
  Second = 'second'
}

export type Format = 'ampm' | '24hr'
export type Period = 'am' | 'pm'
export type AllowFunction = (val: number) => boolean

interface Allowed {
  hours: AllowFunction | number[]
  minutes: AllowFunction | number[]
  seconds: AllowFunction | number[]
}

export interface Time {
  hour: number | null
  minute: number | null
  second: number | null
}

export function convert24to12 (hour: number) {
  return hour ? ((hour - 1) % 12 + 1) : 12
}

export function convert12to24 (hour: number, period: Period) {
  return hour % 12 + (period === 'pm' ? 12 : 0)
}

export function parseTime (value: string | null | Date): [Time, Period] {
  let hour, minute, second, period

  if (value == null || value === '') {
    hour = null
    minute = null
    second = null
    period = 'am'
  } else if (value instanceof Date) {
    hour = value.getHours()
    minute = value.getMinutes()
    second = value.getSeconds()
    period = hour > 12 ? 'pm' : 'am'
  } else {
    const [, h, m, , s, p] = value.trim().toLowerCase().match(/^(\d+):(\d+)(:(\d+))?([ap]m)?$/) || new Array(6)

    hour = p ? convert12to24(parseInt(h, 10), p as Period) : parseInt(h, 10)
    minute = parseInt(m, 10)
    second = parseInt(s || 0, 10)
    period = p || (hour >= 0 && hour < 12 ? 'am' : 'pm')
  }

  return [{ hour, minute, second }, period as Period]
}

export interface VTimeScopedProps {
  allowed: {
    hour: Function
    minute: Function
    second: Function
  }
  format: Format
  time: Time
  period: Period
  selectMode: SelectMode
  setPeriod: Function
  setTime: Function
  setSelectMode: Function
  useSeconds: boolean
}

export default Vue.extend({
  name: 'v-time',

  props: {
    allowed: {
      type: Object,
      default: () => ({ hours: () => true, minutes: () => true, seconds: () => true })
    } as PropValidator<Allowed>,
    format: {
      type: String,
      default: 'ampm',
      validator (val) {
        return ['ampm', '24hr'].includes(val)
      }
    } as PropValidator<Format>,
    min: String,
    max: String,
    value: {
      type: [String, Date],
      default: () => null
    },
    useSeconds: Boolean
  },

  data () {
    const [internalTime, period] = parseTime(this.value)

    return {
      period,
      selectMode: SelectMode.Hour,
      internalTime
    }
  },

  computed: {
    isAllowedHourCb (): AllowFunction {
      let cb: AllowFunction

      if (!this.allowed.hours) {
        cb = () => true
      } else if (this.allowed.hours instanceof Array) {
        cb = (val: number) => (this.allowed.hours as number[]).includes(val)
      } else {
        cb = this.allowed.hours
      }

      if (!this.min && !this.max) return cb

      const minHour = this.min ? Number(this.min.split(':')[0]) : 0
      const maxHour = this.max ? Number(this.max.split(':')[0]) : 23

      return (val: number) => {
        return val >= minHour * 1 &&
          val <= maxHour * 1 &&
          (!cb || cb(val))
      }
    },
    isAllowedMinuteCb (): AllowFunction {
      let cb: AllowFunction

      const isHourAllowed = !this.isAllowedHourCb || this.internalTime.hour === null || this.isAllowedHourCb(this.internalTime.hour)
      if (!this.allowed.minutes) {
        cb = () => true
      } else if (this.allowed.minutes instanceof Array) {
        cb = (val: number) => (this.allowed.minutes as number[]).includes(val)
      } else {
        cb = this.allowed.minutes
      }

      if (!this.min && !this.max) {
        return isHourAllowed ? cb : () => false
      }

      const [minHour, minMinute] = this.min ? this.min.split(':').map(Number) : [0, 0]
      const [maxHour, maxMinute] = this.max ? this.max.split(':').map(Number) : [23, 59]
      const minTime = minHour * 60 + minMinute * 1
      const maxTime = maxHour * 60 + maxMinute * 1

      return (val: number) => {
        const time = 60 * this.internalTime.hour! + val
        return time >= minTime &&
          time <= maxTime &&
          isHourAllowed &&
          (!cb || cb(val))
      }
    },
    isAllowedSecondCb (): AllowFunction {
      let cb: AllowFunction

      const isHourAllowed = !this.isAllowedHourCb || this.internalTime.hour === null || this.isAllowedHourCb(this.internalTime.hour)
      const isMinuteAllowed = isHourAllowed &&
        (!this.isAllowedMinuteCb ||
          this.internalTime.minute === null ||
          this.isAllowedMinuteCb(this.internalTime.minute)
        )

      if (!this.allowed.seconds) {
        cb = () => true
      } else if (this.allowed.seconds instanceof Array) {
        cb = (val: number) => (this.allowed.seconds as number[]).includes(val)
      } else {
        cb = this.allowed.seconds
      }

      if (!this.min && !this.max) {
        return isMinuteAllowed ? cb : () => false
      }

      const [minHour, minMinute, minSecond] = this.min ? this.min.split(':').map(Number) : [0, 0, 0]
      const [maxHour, maxMinute, maxSecond] = this.max ? this.max.split(':').map(Number) : [23, 59, 59]
      const minTime = minHour * 3600 + minMinute * 60 + (minSecond || 0) * 1
      const maxTime = maxHour * 3600 + maxMinute * 60 + (maxSecond || 0) * 1

      return (val: number) => {
        const time = 3600 * this.internalTime.hour! + 60 * this.internalTime.minute! + val
        return time >= minTime &&
          time <= maxTime &&
          isMinuteAllowed &&
          (!cb || cb(val))
      }
    },
    scopedSlotProps (): VTimeScopedProps {
      return {
        allowed: {
          hour: this.isAllowedHourCb,
          minute: this.isAllowedMinuteCb,
          second: this.isAllowedSecondCb
        },
        format: this.format,
        time: this.internalTime,
        period: this.period,
        selectMode: this.selectMode,
        setPeriod: this.setPeriod,
        setTime: this.setTime,
        setSelectMode: this.setSelectMode,
        useSeconds: this.useSeconds
      }
    },
    timeAsString (): string | null {
      const { hour, minute, second } = this.internalTime
      if (hour != null && minute != null && (!this.useSeconds || second != null)) {
        return `${pad(hour)}:${pad(minute)}` + (this.useSeconds ? `:${pad(second!)}` : '')
      }

      return null
    }
  },

  watch: {
    value (value: string | null | Date) {
      const [time, period] = parseTime(value)
      this.internalTime = time
      this.period = period
    },
    timeAsString (v: string | null) {
      if (v != null) this.$emit('input', v)
    },
    period (v: Period) {
      this.$emit('update:period', v)
    },
    selectMode (v: SelectMode) {
      this.$emit('update:selectMode', v)
    }
  },

  methods: {
    setPeriod (p: Period) {
      this.period = p

      // If hour is set, and we change period
      // then we need to update hour since it
      // is always stored in a 24 hour format
      if (this.internalTime.hour) {
        this.internalTime.hour = this.period === 'am' ? this.internalTime.hour - 12 : this.internalTime.hour + 12
      }
    },
    setTime (t: Time) { this.internalTime = t },
    setSelectMode (m: SelectMode) { this.selectMode = m }
  },

  render (): VNode {
    return this.$scopedSlots.default!(this.scopedSlotProps) as any
  }
})
