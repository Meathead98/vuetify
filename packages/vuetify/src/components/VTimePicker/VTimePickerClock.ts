import './VTimePickerClock.sass'

// Mixins
import Colorable from '../../mixins/colorable'
import Themeable from '../../mixins/themeable'

// Helpers
import { pad } from '../VDatePicker/util'
import { deepEqual, convertToUnit } from '../../util/helpers'
import { genPickerButton } from '../VPicker/VPicker'

// Types
import mixins, { ExtractVue } from '../../util/mixins'
import Vue, { VNode } from 'vue'
import { SelectMode, Time, convert24to12, AllowFunction, Format, convert12to24, Period } from './VTime'
import { PropValidator } from 'vue/types/options'

interface Point {
  x: number
  y: number
}

interface options extends Vue {
  $refs: {
    clock: HTMLElement
    innerClock: HTMLElement
  }
}

function euclidean (p0: Point, p1: Point) {
  const dx = p1.x - p0.x
  const dy = p1.y - p0.y

  return Math.sqrt(dx * dx + dy * dy)
}

function angle (center: Point, p1: Point) {
  const value = 2 * Math.atan2(p1.y - center.y - euclidean(center, p1), p1.x - center.x)
  return Math.abs(value * 180 / Math.PI)
}

interface Allowed {
  hour: AllowFunction
  minute: AllowFunction
  second: AllowFunction
}

export default mixins<options &
/* eslint-disable indent */
  ExtractVue<[
    typeof Colorable,
    typeof Themeable
  ]>
/* eslint-enable indent */
>(
  Colorable,
  Themeable
/* @vue/component */
).extend({
  name: 'v-time-picker-clock',

  inheritAttrs: false,

  props: {
    allowed: {
      type: Object,
      default: () => ({ hour: () => true, minute: () => true, second: () => true })
    } as PropValidator<Allowed>,
    disabled: Boolean,
    period: String as PropValidator<Period>,
    readonly: Boolean,
    format: String as PropValidator<Format>,
    rotate: {
      type: Number,
      default: 0
    },
    scrollable: Boolean,
    selectMode: String as PropValidator<SelectMode>,
    showAmPm: Boolean,
    useSeconds: Boolean,
    size: [Number, String],
    time: Object as PropValidator<Time>
  },

  data () {
    return {
      internalSelectMode: this.selectMode,
      isDragging: false,
      valueOnMouseUp: null as number | null,
      min: 0
    }
  },

  computed: {
    isAmPm (): boolean {
      return this.format === 'ampm'
    },
    isHourMode (): boolean {
      return this.internalSelectMode === SelectMode.Hour
    },
    isTwoCircles (): boolean {
      return this.isHourMode && !this.isAmPm
    },
    step (): number {
      return this.isHourMode ? 1 : 5
    },
    max (): number {
      return this.isHourMode ? (this.isAmPm ? 11 : 23) : 59
    },
    unitCount (): number {
      return this.max - this.min + 1
    },
    degreesPerUnit (): number {
      return 360 / this.unitsPerCircle
    },
    degrees (): number {
      return this.degreesPerUnit * Math.PI / 180
    },
    value (): number | null {
      if (!this.time) return null

      let value = null
      switch (this.internalSelectMode) {
        case SelectMode.Hour: value = this.time.hour; break
        case SelectMode.Minute: value = this.time.minute; break
        case SelectMode.Second: value = this.time.second; break
      }

      return value
    },
    innerRadiusScale (): number {
      return 0.62
    },
    unitsPerCircle (): number {
      return this.isTwoCircles ? (this.unitCount / 2) : this.unitCount
    }
  },

  watch: {
    selectMode (v: SelectMode) {
      this.internalSelectMode = v
    }
  },

  methods: {
    cycleMode () {
      let selectMode = this.internalSelectMode

      if (selectMode === SelectMode.Hour) {
        selectMode = SelectMode.Minute
      } else if (this.useSeconds && selectMode === SelectMode.Minute) {
        selectMode = SelectMode.Second
      }

      this.internalSelectMode = selectMode
      this.$emit('update:selectMode', selectMode)
    },
    wheel (e: WheelEvent) {
      e.preventDefault()

      const delta = Math.sign(-e.deltaY || 1)
      let value = this.value || 0
      do {
        value = value + delta
        value = (value - this.min + this.unitCount) % this.unitCount + this.min
      } while (!this.isAllowed(value) && value !== this.value)

      if (value !== this.value) {
        this.update(value)
      }
    },
    isInner (value: number): boolean {
      return this.isTwoCircles && (value - this.min >= this.unitsPerCircle)
    },
    handScale (value: number): number {
      return this.isInner(value) ? this.innerRadiusScale : 1
    },
    isAllowed (value: number): boolean {
      switch (this.internalSelectMode) {
        case SelectMode.Hour: return !this.allowed.hour || this.allowed.hour(value)
        case SelectMode.Minute: return !this.allowed.minute || this.allowed.minute(value)
        case SelectMode.Second: return !this.allowed.second || this.allowed.second(value)
      }
    },
    genValues () {
      const children: VNode[] = []

      for (let value = this.min; value <= this.max; value = value + this.step) {
        const compareValue = this.isHourMode && this.isAmPm ? convert24to12(value) : value
        const currentValue = this.isHourMode && this.isAmPm ? convert24to12(this.value || 0) : this.value || 0
        const color = this.value !== null && compareValue === currentValue && (this.color || 'accent')
        const displayValue = !this.isHourMode ? pad(value, 2) : compareValue

        children.push(this.$createElement('span', this.setBackgroundColor(color, {
          staticClass: 'v-time-picker-clock__item',
          'class': {
            'v-time-picker-clock__item--active': compareValue === currentValue,
            'v-time-picker-clock__item--disabled': this.disabled || !this.isAllowed(value)
          },
          style: this.getTransform(value),
          domProps: { innerHTML: `<span>${displayValue}</span>` }
        })))
      }

      return children
    },
    genHand (): VNode {
      const value = this.value || 0
      const scale = `scaleY(${this.handScale(value)})`
      const angle = this.rotate + this.degreesPerUnit * (value - this.min)
      const color = (this.value !== null) && (this.color || 'accent')
      return this.$createElement('div', this.setBackgroundColor(color, {
        staticClass: 'v-time-picker-clock__hand',
        'class': {
          'v-time-picker-clock__hand--inner': this.isInner(value)
        },
        style: {
          transform: `rotate(${angle}deg) ${scale}`
        }
      }))
    },
    getTransform (i: number) {
      const { x, y } = this.getPosition(i)
      return {
        left: `${50 + x * 50}%`,
        top: `${50 + y * 50}%`
      }
    },
    getPosition (value: number): Point {
      const rotateRadians = this.rotate * Math.PI / 180
      return {
        x: Math.sin((value - this.min) * this.degrees + rotateRadians) * this.handScale(value),
        y: -Math.cos((value - this.min) * this.degrees + rotateRadians) * this.handScale(value)
      }
    },
    onMouseDown (e: MouseEvent | TouchEvent) {
      e.preventDefault()

      this.valueOnMouseUp = null
      this.isDragging = true
      this.onDragMove(e)
    },
    onMouseUp () {
      this.isDragging = false
      if (this.valueOnMouseUp !== null && this.isAllowed(this.valueOnMouseUp)) {
        this.update(this.valueOnMouseUp)
        this.$emit(`click:${this.internalSelectMode}`, this.valueOnMouseUp)
        this.cycleMode()
      }
    },
    onDragMove (e: MouseEvent | TouchEvent) {
      e.preventDefault()
      if (!this.isDragging && e.type !== 'click') return

      const { width, top, left } = this.$refs.clock.getBoundingClientRect()
      const { width: innerWidth } = this.$refs.innerClock.getBoundingClientRect()
      const { clientX, clientY } = 'touches' in e ? e.touches[0] : e
      const center = { x: width / 2, y: -width / 2 }
      const coords = { x: clientX - left, y: top - clientY }
      const handAngle = Math.round(angle(center, coords) - this.rotate + 360) % 360
      const insideClick = this.isTwoCircles && euclidean(center, coords) < (innerWidth + innerWidth * this.innerRadiusScale) / 4
      const value = (
        Math.round(handAngle / this.degreesPerUnit) +
        (insideClick ? this.unitsPerCircle : 0)
      ) % this.unitCount + this.min

      // Necessary to fix edge case when selecting left part of the value(s) at 12 o'clock
      let newValue: number
      if (handAngle >= (360 - this.degreesPerUnit / 2)) {
        newValue = insideClick ? this.max - this.unitsPerCircle + 1 : this.min
      } else {
        newValue = value
      }

      if (this.isAllowed(value)) {
        this.valueOnMouseUp = newValue
        this.update(newValue)
      }
    },
    update (value: number) {
      const time = { ...this.time }

      switch (this.internalSelectMode) {
        case SelectMode.Hour: time.hour = this.isAmPm ? convert12to24(value, this.period) : value; break
        case SelectMode.Minute: time.minute = value; break
        case SelectMode.Second: time.second = value; break
      }

      if (deepEqual(this.time, time)) return

      this.$emit('update:time', time)
    },
    genClock () {
      let listeners = {}

      if (!this.readonly && !this.disabled) {
        listeners = Object.assign(listeners, {
          mousedown: this.onMouseDown,
          mouseup: this.onMouseUp,
          mouseleave: () => (this.isDragging && this.onMouseUp()),
          touchstart: this.onMouseDown,
          touchend: this.onMouseUp,
          mousemove: this.onDragMove,
          touchmove: this.onDragMove
        })

        if (this.scrollable) {
          listeners = Object.assign(listeners, {
            wheel: this.wheel
          })
        }
      }

      const data = {
        staticClass: 'v-time-picker-clock__outer',
        on: listeners,
        ref: 'clock'
      }

      return this.$createElement('div', data, [
        this.$createElement('div', {
          staticClass: 'v-time-picker-clock__inner',
          ref: 'innerClock'
        }, [
          this.genHand(),
          this.genValues()
        ])
      ])
    },
    genClockAmPm () {
      return this.$createElement('div', this.setTextColor(this.color || 'primary', {
        staticClass: 'v-time-picker-clock__ampm'
      }), [
        genPickerButton(
          this.$createElement,
          'AM',
          () => this.$emit('update:period', 'am'),
          this.period === 'am',
          this.disabled || this.readonly
        ),
        genPickerButton(
          this.$createElement,
          'PM',
          () => this.$emit('update:period', 'pm'),
          this.period === 'pm',
          this.disabled || this.readonly
        )
      ])
    }
  },

  render (h): VNode {
    return this.$createElement('div', {
      staticClass: 'v-time-picker-clock',
      class: {
        'v-time-picker-clock--indeterminate': this.value == null,
        ...this.themeClasses
      },
      style: {
        width: convertToUnit(this.size),
        height: convertToUnit(this.size)
      }
    }, [
      this.isAmPm && this.showAmPm && this.genClockAmPm(),
      this.genClock()
    ])
  }
})
