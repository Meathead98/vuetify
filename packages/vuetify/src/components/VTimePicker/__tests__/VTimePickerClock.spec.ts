import VTimePickerClock from '../VTimePickerClock'
import { touch, mouse } from '../../../../test'
import { mount, MountOptions, Wrapper } from '@vue/test-utils'
import { SelectMode } from '../VTime'

export function createBoundingRect (wrapper, outerRect = { width: 270, top: 10, left: 10 }, innerRect = { width: 216 }) {
  wrapper.vm.$refs.clock.getBoundingClientRect = () => {
    return outerRect
  }

  wrapper.vm.$refs.innerClock.getBoundingClientRect = () => {
    return innerRect
  }
}

describe('VTimePickerClock.ts', () => {
  (window as any).TouchEvent = Event

  type Instance = InstanceType<typeof VTimePickerClock>
  let mountFunction: (options?: MountOptions<Instance>) => Wrapper<Instance>
  beforeEach(() => {
    mountFunction = (options?: MountOptions<Instance>) => {
      return mount(VTimePickerClock, options)
    }
  })

  it('should render hours in 24hr format', () => {
    const wrapper = mountFunction({
      propsData: {
        selectMode: SelectMode.Hour,
        format: '24hr'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render hours in ampm format', async () => {
    const wrapper = mountFunction({
      propsData: {
        selectMode: SelectMode.Hour,
        format: 'ampm'
      }
    })

    await wrapper.vm.$nextTick()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should emit input event on wheel if scrollable', async () => {
    const time = jest.fn()
    const wrapper = mountFunction({
      propsData: {
        time: {
          hour: 5,
          minute: 4,
          second: 3
        },
        selectMode: SelectMode.Hour
      },
      listeners: {
        'update:time': time
      }
    })

    const clock = wrapper.find('.v-time-picker-clock__outer')
    clock.trigger('wheel')
    await wrapper.vm.$nextTick()

    expect(time).not.toHaveBeenCalled()

    wrapper.setProps({
      scrollable: true
    })
    await wrapper.vm.$nextTick()

    clock.trigger('wheel')
    await wrapper.vm.$nextTick()

    expect(time).toHaveBeenCalledWith({ hour: 6, minute: 4, second: 3 })
  })

  it('should not emit input event on wheel if scrollable and disabled or readonly', async () => {
    const time = jest.fn()
    const wrapper = mountFunction({
      propsData: {
        time: {
          hour: 5,
          minute: 4,
          second: 3
        },
        selectMode: SelectMode.Hour,
        scrollable: true,
        disabled: true
      },
      listeners: {
        'update:time': time
      }
    })

    const clock = wrapper.find('.v-time-picker-clock__outer')
    clock.trigger('wheel')
    await wrapper.vm.$nextTick()

    wrapper.setProps({
      disabled: false,
      readonly: true
    })
    await wrapper.vm.$nextTick()

    clock.trigger('wheel')
    await wrapper.vm.$nextTick()

    expect(time).not.toHaveBeenCalled()
  })

  it('should not emit input event when clicked disabled value (#5897)', async () => {
    const input = jest.fn()
    const wrapper = mountFunction({
      propsData: {
        allowed: {
          minute: m => m >= 30
        },
        selectMode: SelectMode.Minute
      },
      listeners: {
        'update:time': input
      }
    })

    createBoundingRect(wrapper)

    const clock = wrapper.find('.v-time-picker-clock__outer')

    mouse(clock).mousedown(270, 270).mouseup()
    await wrapper.vm.$nextTick()

    expect(input).not.toHaveBeenCalled()

    mouse(clock).mousedown(10, 150).mouseup()
    await wrapper.vm.$nextTick()

    expect(input).toHaveBeenCalledWith({ minute: 45 })
  })

  it('should change with touch move', async () => {
    const input = jest.fn()
    const wrapper = mount(VTimePickerClock, {
      propsData: {
        size: '290',
        selectMode: SelectMode.Minute
      },
      listeners: {
        'update:time': input
      }
    })

    createBoundingRect(wrapper)

    const clock = wrapper.find('.v-time-picker-clock__outer')

    const finger = touch(clock).start(255, 145)
    await wrapper.vm.$nextTick()
    expect(input).toHaveBeenCalledWith({ minute: 15 })

    finger.move(200, 240)
    await wrapper.vm.$nextTick()
    expect(input).toHaveBeenCalledWith({ minute: 25 })

    finger.move(66, 66)
    await wrapper.vm.$nextTick()
    expect(input).toHaveBeenCalledWith({ minute: 53 })

    // edge case, to the left of 0
    finger.move(137, 41)
    await wrapper.vm.$nextTick()
    expect(input).toHaveBeenCalledWith({ minute: 59 })
  })
})
