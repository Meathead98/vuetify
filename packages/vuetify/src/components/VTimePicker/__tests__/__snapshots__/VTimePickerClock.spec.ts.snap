// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VTimePickerClock.ts should render hours in 24hr format 1`] = `

<div class="v-time-picker-clock v-time-picker-clock--indeterminate theme--light">
  <div class="v-time-picker-clock__outer">
    <div class="v-time-picker-clock__inner">
      <div class="v-time-picker-clock__hand"
           style="transform: rotate(0deg) scaleY(1);"
      >
      </div>
      <span class="v-time-picker-clock__item v-time-picker-clock__item--active"
            style="left: 50%; top: 0%;"
      >
        <span>
          0
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 75%; top: 6.698729810778062%;"
      >
        <span>
          1
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 93.30127018922192%; top: 24.999999999999993%;"
      >
        <span>
          2
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 100%; top: 50%;"
      >
        <span>
          3
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 93.30127018922194%; top: 74.99999999999999%;"
      >
        <span>
          4
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 75.00000000000001%; top: 93.30127018922192%;"
      >
        <span>
          5
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 50.00000000000001%; top: 100%;"
      >
        <span>
          6
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 25.000000000000014%; top: 93.30127018922194%;"
      >
        <span>
          7
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 6.698729810778076%; top: 75.00000000000003%;"
      >
        <span>
          8
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 0%; top: 50.00000000000001%;"
      >
        <span>
          9
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 6.6987298107780475%; top: 25.000000000000032%;"
      >
        <span>
          10
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 24.99999999999998%; top: 6.698729810778083%;"
      >
        <span>
          11
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 49.99999999999999%; top: 19%;"
      >
        <span>
          12
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 65.49999999999997%; top: 23.153212482682388%;"
      >
        <span>
          13
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 76.84678751731758%; top: 34.499999999999986%;"
      >
        <span>
          14
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 81%; top: 49.999999999999964%;"
      >
        <span>
          15
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 76.84678751731761%; top: 65.49999999999997%;"
      >
        <span>
          16
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 65.50000000000004%; top: 76.84678751731758%;"
      >
        <span>
          17
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 50.000000000000014%; top: 81%;"
      >
        <span>
          18
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 34.50000000000003%; top: 76.84678751731761%;"
      >
        <span>
          19
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 23.153212482682427%; top: 65.50000000000004%;"
      >
        <span>
          20
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 19%; top: 50.000000000000014%;"
      >
        <span>
          21
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 23.153212482682385%; top: 34.50000000000003%;"
      >
        <span>
          22
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 34.49999999999996%; top: 23.153212482682427%;"
      >
        <span>
          23
        </span>
      </span>
    </div>
  </div>
</div>

`;

exports[`VTimePickerClock.ts should render hours in ampm format 1`] = `

<div class="v-time-picker-clock v-time-picker-clock--indeterminate theme--light">
  <div class="v-time-picker-clock__outer">
    <div class="v-time-picker-clock__inner">
      <div class="v-time-picker-clock__hand"
           style="transform: rotate(0deg) scaleY(1);"
      >
      </div>
      <span class="v-time-picker-clock__item v-time-picker-clock__item--active"
            style="left: 50%; top: 0%;"
      >
        <span>
          12
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 75%; top: 6.698729810778062%;"
      >
        <span>
          1
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 93.30127018922192%; top: 24.999999999999993%;"
      >
        <span>
          2
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 100%; top: 50%;"
      >
        <span>
          3
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 93.30127018922194%; top: 74.99999999999999%;"
      >
        <span>
          4
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 75.00000000000001%; top: 93.30127018922192%;"
      >
        <span>
          5
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 50.00000000000001%; top: 100%;"
      >
        <span>
          6
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 25.000000000000014%; top: 93.30127018922194%;"
      >
        <span>
          7
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 6.698729810778076%; top: 75.00000000000003%;"
      >
        <span>
          8
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 0%; top: 50.00000000000001%;"
      >
        <span>
          9
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 6.6987298107780475%; top: 25.000000000000032%;"
      >
        <span>
          10
        </span>
      </span>
      <span class="v-time-picker-clock__item"
            style="left: 24.99999999999998%; top: 6.698729810778083%;"
      >
        <span>
          11
        </span>
      </span>
    </div>
  </div>
</div>

`;
